src/mapgen_valleys.cpp

```diff
@@ -897,101 +783,150 @@ void MapgenValleys::dustTopNodes()
 }
 
 
-void MapgenValleys::generateSimpleCaves(s16 max_stone_y)
+void MapgenValleys::generateCaves(s16 max_stone_y)
 {
+	if (max_stone_y < node_min.Y)
+		return;
+
 	PseudoRandom ps(blockseed + 72202);
 
 	MapNode n_air(CONTENT_AIR);
-	MapNode n_dirt(c_dirt);
 	MapNode n_lava(c_lava_source);
 	MapNode n_water(c_river_water_source);
 
 	v3s16 em = vm->m_area.getExtent();
 
-	s16 base_water_chance = 0;
-	if (water_features < 11)
-		base_water_chance = ceil(MAX_MAP_GENERATION_LIMIT / (water_features * 1000));
-
-	if (max_stone_y >= node_min.Y) {
-		u32 index_2d = 0;
-		u32 index_3d = 0;
-		for (s16 z = node_min.Z; z <= node_max.Z; z++)
-		for (s16 x = node_min.X; x <= node_max.X; x++, index_2d++) {
-			bool air_above = false;
-			//bool underground = false;
-			u32 index_data = vm->m_area.index(x, node_max.Y + 1, z);
-
-			index_3d = (z - node_min.Z) * zstride + (csize.Y + 1) * ystride + (x - node_min.X);
-
-			// Dig caves on down loop to check for air above.
-			for (s16 y = node_max.Y + 1;
-					y >= node_min.Y - 1;
-					y--, index_3d -= ystride, vm->m_area.add_y(em, index_data, -1)) {
-				float terrain = noise_terrain_height->result[index_2d];
-
-				// Saves some time and prevents removing above ground nodes.
-				if (y > terrain + 1) {
-					air_above = true;
-					continue;
-				}
+	// Cave blend distance near YMIN, YMAX
+	const float massive_cave_blend = 128.f;
+	// noise threshold for massive caves
+	const float massive_cave_threshold = 0.6f;
+	// mct: 1 = small rare caves, 0.5 1/3rd ground volume, 0 = 1/2 ground volume.
 
-				content_t c = vm->m_data[index_data].getContent();
-				bool n1 = (fabs(noise_simple_caves_1->result[index_3d]) < 0.07f);
-				bool n2 = (fabs(noise_simple_caves_2->result[index_3d]) < 0.07f);
+	float yblmin = -map_gen_limit + massive_cave_blend * 1.5f;
+	float yblmax = massive_cave_depth - massive_cave_blend * 1.5f;
+	bool made_a_big_one = false;
+	float tcave_cache[csize.Y + 2];
```

Comment:
VLAs are not allowed in minetest.  This will cause compilation to fail on some platforms.

Please consider making tcave_cache a float * member variable instead, and allocate the cache on mapgen startup using new[].